sceptre/resolvers/__init__.py

```diff
@@ -17,37 +17,169 @@
+        instance's setup method is called.
+        """
+        self.setup()
+
+        def setup_nested(containing_list_or_dict, key, obj: Resolver):
+            obj._recursively_setup()
+
+        _call_func_on_values(setup_nested, self._argument, Resolver)
+
+    def _recursively_clone(self: Self, stack: "stack.Stack") -> Self:
+        """Recursively clones the instance and its arguments.
+
+        The returned instance will have an identical argument that is a different memory reference,
+        so that instances inherited from a stack group and applied across multiple stacks are
+        independent of each other.
+
+        Furthermore, all nested resolvers in this resolver's argument will also be cloned to ensure
+        they themselves are also independent and fully configured for the current stack.
+        """
+
+        def recursively_clone_arguments(obj):
+            if isinstance(obj, Resolver):
+                return obj._recursively_clone(stack)
+            if isinstance(obj, list):
+                return [recursively_clone_arguments(item) for item in obj]
+            elif isinstance(obj, dict):
+                return {
+                    key: recursively_clone_arguments(val) for key, val in obj.items()
+                }
+            return obj
+
+        argument = recursively_clone_arguments(self._argument)
+        clone = type(self)(argument, stack)
+        return clone
+
+    def clone_for_stack(self: Self, stack: "stack.Stack") -> Self:
+        """
+        Obtains a clone of the current object, setup and ready for use for a given Stack instance.
+        """
+        clone = self._recursively_clone(stack)
+        clone._recursively_setup()
+        return clone
+
     def setup(self):
         """
-        This method is called at during stack initialisation.
+        This method is called when the object is connected to a Stack instance.
         Implementation of this method in subclasses can be used to do any
         initial setup of the object.
         """
         pass  # pragma: no cover
 
+    def __repr__(self) -> str:
+        """Returns a string representation of the resolver.
+
+        This is mostly used for resolver placeholders. In cases where we cannot resolve a resolver
+        YET, such as when we're generating a template or diff and there's a dependency on a stack
+        output of a stack that hasn't been deployed yet, placeholders need to render the resolver
+        in a way that is useful.
+
+        We use self._argument instead of self.argument because if there are resolvers nested in this
+        resolver's argument and one of those cannot be resolved, we'll need those resolvers to also
+        be converted to useful placeholder values.
+        """
+        as_str = f"!{self.__class__.__name__}"
+        if self._argument is not None:
+            as_str += f"({self._argument})"
+
+        return as_str
+
+
+class Resolver(CustomYamlTagBase, metaclass=abc.ABCMeta):
```

Comment:
I've moved most of the functionality that _was_ in the Resolver class into the CustomYamlTagBase, which is now shared by both the Resolver and Hook classes.